instant-distance-py/src/lib.rs

```diff
@@ -414,73 +413,42 @@ impl Neighbor {
     }
 }
 
-#[repr(align(32))]
 #[derive(Clone, Deserialize, Serialize)]
-struct FloatArray(#[serde(with = "BigArray")] [f32; DIMENSIONS]);
+struct FloatArray<M> {
+    array: Vec<f32>,
+    phantom: PhantomData<M>,
+}
 
-impl FloatArray {
+impl<M: Metric> FloatArray<M> {
     fn try_from_pylist(list: &PyList) -> Result<Vec<Self>, PyErr> {
         list.into_iter().map(FloatArray::try_from).collect()
     }
 }
 
-impl TryFrom<&PyAny> for FloatArray {
+impl<M: Metric> From<Vec<f32>> for FloatArray<M> {
+    fn from(array: Vec<f32>) -> Self {
+        Self {
+            array,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<M: Metric> TryFrom<&PyAny> for FloatArray<M> {
     type Error = PyErr;
 
     fn try_from(value: &PyAny) -> Result<Self, Self::Error> {
-        let mut new = FloatArray([0.0; DIMENSIONS]);
-        for (i, val) in value.iter()?.enumerate() {
-            match i >= DIMENSIONS {
-                true => return Err(PyTypeError::new_err("point array too long")),
-                false => new.0[i] = val?.extract::<f32>()?,
-            }
-        }
-        Ok(new)
+        let array: Vec<f32> = value
+            .iter()?
+            .map(|val| val.and_then(|v| v.extract::<f32>()))
+            .collect::<Result<_, _>>()?;
+        Ok(Self::from(array))
     }
 }
 
-impl Point for FloatArray {
```

Comment:
So this commit still mixes the adding/changing of a lot of code with moving it around. Usually I'd request a separate commit that just refactors the code so that the implementation here is moved into its new place (introducing the `Metric` trait) and the code around it is left more or less unchanged, so that it's easier to review these things separately. So then I'd probably end up with:

* Implement `Metric` trait and parametrize `FloatArray` with it
* Add new `Metric` types and impls
* Change `FloatArray` to `FloatVector`